File: tools/refactor_imports.py

import os
import re

# Map alias prefixes to source directories (relative to project root)
ALIAS_MAP = {
    "@core": "src/core",
    "#core": "src/core",
    "@logic": "src/logic",
    "#logic": "src/logic",
    "@data": "src/data",
    "#data": "src/data",
    "@ui": "src/ui",
    "#ui": "src/ui",
    "@helpers": "src/helpers",
    "#helpers": "src/helpers",
    "gameLogic": "src/logic/index.js", # handling specialized path
}

def resolve_alias(import_path, current_file_path):
    # import_path example: "@core/utils.js"
    # current_file_path example: "src/logic/effects/ops/draw.ts"

    for alias, target_dir in ALIAS_MAP.items():
        if import_path.startswith(alias):
            # Remove alias part
            remainder = import_path[len(alias):]
            if remainder.startswith("/"):
                remainder = remainder[1:]
            
            # Construct absolute target path (assuming project root is cwd)
            # handle special "gameLogic" case which maps to a file, not just a dir prefix?
            if alias == "gameLogic":
                 target_abs = os.path.abspath(target_dir)
            else:
                 target_abs = os.path.abspath(os.path.join(target_dir, remainder))

            # current directory
            current_dir = os.path.dirname(os.path.abspath(current_file_path))
            
            # Calculate relative path
            rel_path = os.path.relpath(target_abs, current_dir)
            
            # Formatting: ensure ./ if it's in same dir or subdir, replacing \ with /
            rel_path = rel_path.replace("\\", "/")
            if not rel_path.startswith(".") and not rel_path.startswith("/"):
                rel_path = "./" + rel_path
            
            return rel_path
    
    return None

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    new_lines = []
    modified = False
    
    # Regex for import/export ... from '...'
    # Matches: import ... from "..." or export ... from "..."
    # Capture group 2 is the quote type, group 3 is the path
    import_regex = re.compile(r'^(import|export)\s+.*?from\s+([\'"])(.*?)([\'"]);?$')
    # Regex for side-effect import: import "..."
    side_effect_regex = re.compile(r'^import\s+([\'"])(.*?)([\'"]);?$')
    # Regex for dynamic import: import("...")
    dynamic_import_regex = re.compile(r'import\(([\'"])(.*?)([\'"])\)')

    for line in lines:
        new_line = line
        
        # Check standard import/export
        match = import_regex.match(line)
        if match:
            quote = match.group(2)
            path = match.group(3)
            new_path = resolve_alias(path, filepath)
            if new_path:
                new_line = line.replace(f"{quote}{path}{quote}", f"{quote}{new_path}{quote}")
                modified = True
        
        # Check side effect import
        elif side_effect_regex.match(line):
            match = side_effect_regex.match(line)
            quote = match.group(1)
            path = match.group(2)
            new_path = resolve_alias(path, filepath)
            if new_path:
                 new_line = line.replace(f"{quote}{path}{quote}", f"{quote}{new_path}{quote}")
                 modified = True

        # Check dynamic import (naive replace, assumes one per line or simple structure)
        elif "import(" in line:
            match = dynamic_import_regex.search(line)
            if match:
                quote = match.group(1)
                path = match.group(2)
                new_path = resolve_alias(path, filepath)
                if new_path:
                    new_line = line.replace(f"{quote}{path}{quote}", f"{quote}{new_path}{quote}")
                    modified = True

        new_lines.append(new_line)

    if modified:
        print(f"Modifying {filepath}")
        with open(filepath, 'w', encoding='utf-8') as f:
            f.writelines(new_lines)

def main():
    target_exts = ['.ts', '.tsx', '.js', '.jsx']
    for root, dirs, files in os.walk("src"):
        for file in files:
            if any(file.endswith(ext) for ext in target_exts):
                process_file(os.path.join(root, file))
    
    # Also process tests
    for root, dirs, files in os.walk("tests"):
        for file in files:
             if any(file.endswith(ext) for ext in target_exts):
                process_file(os.path.join(root, file))

if __name__ == "__main__":
    main()

File: tools/refactor_imports.mjs

import { Project, SyntaxKind } from "ts-morph";
import path from "path";
import fs from "fs";

// Initialize project based on tsconfig.json to leverage path mapping resolution
const project = new Project({
    tsConfigFilePath: "tsconfig.json",
    skipAddingFilesFromTsConfig: false,
});

console.log("Files found in project: " + project.getSourceFiles().length);

// Also need to handle package.json '#' imports if they aren't covered by tsconfig
// However, tsconfig usually has the source-of-truth for the build.
// The user noted: "@core/* -> src/core/*" etc.
// ts-morph using tsconfig should handle resolving "@core/foo" to "src/core/foo.ts" automatically
// IF the resolution finds the file.

let modifications = 0;

const sourceFiles = project.getSourceFiles();

// Helper to determine if a specifier is an alias we want to rewrite
const aliases = ["@core", "@data", "@logic", "@ui", "@helpers", "@ops", "#core", "#logic", "#ui", "#helpers", "#data"];

function isAliased(specifier) {
    return aliases.some(a => specifier.startsWith(a));
}

function processFile(sourceFile) {
    const filePath = sourceFile.getFilePath();
    // Skip dist and node_modules explicitly just in case (though tsconfig exclusion handles most)
    if (filePath.includes("/dist/") || filePath.includes("/node_modules/")) return;

    // Helper to rewrite a specifier
    const rewrite = (specifier, nodeToWarn) => {
        if (!isAliased(specifier)) return null;

        // Try to resolve the module specifier to a source file
        const resolvedSourceFile = nodeToWarn.getModuleSpecifierSourceFile();

        if (resolvedSourceFile) {
            const resolvedPath = resolvedSourceFile.getFilePath();

            // Allow extensionless relative path calculation
            let relativePath = path.relative(path.dirname(filePath), resolvedPath);

            // Normalize separators
            relativePath = relativePath.split(path.sep).join("/");

            // Ensure ./ prefix
            if (!relativePath.startsWith(".")) {
                relativePath = "./" + relativePath;
            }

            // Remove extension (.ts, .tsx, .d.ts, .js)
            // User requested explicit extensionless imports
            const ext = path.extname(relativePath);
            if (ext) {
                relativePath = relativePath.substring(0, relativePath.length - ext.length);
            }

            // Special handling for index files?
            // "src/logic/index.ts" -> resolved as "../logic/index" -> "../logic" is cleaner but "../logic/index" is valid.
            // Keeping "../logic/index" is safer for now unless we want to strip /index
            if (relativePath.endsWith("/index")) {
                relativePath = relativePath.substring(0, relativePath.length - 6);
                if (relativePath === "") relativePath = ".";
            }

            return relativePath;

        } else {
            // Fallback or warning
            // If we can't resolve it via AST (maybe it's a runtime-only alias not in tsconfig properly?), we log
            console.warn(`[WARN] Could not resolve '${specifier}' in ${filePath}`);
            return null;
        }
    }

    // Rewrite ImportDeclarations
    sourceFile.getImportDeclarations().forEach(importDecl => {
        const specifier = importDecl.getModuleSpecifierValue();
        const newSpecifier = rewrite(specifier, importDecl);
        if (newSpecifier) {
            importDecl.setModuleSpecifier(newSpecifier);
            modifications++;
        }
    });

    // Rewrite ExportDeclarations (export * from "...", export { x } from "...")
    sourceFile.getExportDeclarations().forEach(exportDecl => {
        const specifier = exportDecl.getModuleSpecifierValue();
        if (specifier) {
            const newSpecifier = rewrite(specifier, exportDecl);
            if (newSpecifier) {
                exportDecl.setModuleSpecifier(newSpecifier);
                modifications++;
            }
        }
    });

    // Rewrite ImportEqualsDeclarations? (Usually for CommonJS: import x = require('...'))
    // Not expecting much in ESNext modules but good to check
    sourceFile.getImportEqualsDeclarations().forEach(importDecl => {
        const ref = importDecl.getModuleReference();
        if (SyntaxKind.ExternalModuleReference && ref.getKind() === SyntaxKind.ExternalModuleReference) {
            const specifierExpression = ref.getExpression();
            if (specifierExpression && (specifierExpression.getKind() === SyntaxKind.StringLiteral)) {
                const specifier = specifierExpression.getLiteralText();
                // Cannot easily use getModuleSpecifierSourceFile on ImportEquals
                // Skipping for now unless we see errors.
            }
        }
    });

    // Dynamic Imports: import("...")
    // ts-morph wraps these in CallExpression
    const callExpressions = sourceFile.getDescendantsOfKind(SyntaxKind.CallExpression);
    callExpressions.forEach(callExpr => {
        if (callExpr.getExpression().getKind() === SyntaxKind.ImportKeyword) {
            const args = callExpr.getArguments();
            if (args.length > 0 && args[0].getKind() === SyntaxKind.StringLiteral) {
                const stringLiteral = args[0];
                const specifier = stringLiteral.getLiteralText();

                // Manual resolution attempt or leverage TypeChecker?
                // ts-morph's getSymbol() or local resolution might be hard on dynamic import
                // But we can try to resolve it if it matches our alias pattern

                if (isAliased(specifier)) {
                    // Since we can't easily get "ModuleSpecifierSourceFile" from a dynamic import string literal node directly
                    // in the same way, we might have to rely on the fact that if it's an alias we know,
                    // we should try to find it.
                    // However, without the resolver, we are guessing.
                    // But wait, we can ask the TypeChecker to resolve the module specifier?
                    // Or just skip dynamic imports for now?
                    // Let's log it.
                    console.log(`[INFO] Found dynamic import '${specifier}' in ${filePath}. Attempting resolve.`);

                    // Try to use the project's resolution (hacky)
                    // If we are strict, maybe we just log fail for now.
                    // Or we can try to find the file manually? 
                    // Let's mark as manual intervention needed if we see output.
                }
            }
        }
    });
}

sourceFiles.forEach(processFile);

if (modifications > 0) {
    console.log(`Saving ${modifications} changes...`);
    project.saveSync();
    console.log("Done.");
} else {
    console.log("No changes made.");
}
